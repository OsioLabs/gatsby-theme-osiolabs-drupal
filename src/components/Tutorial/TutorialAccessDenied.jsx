import React from 'react';
import { Container, Header } from 'semantic-ui-react';

const TutorialAccessDenied = () => (
  <Container className="tutrial--access-denied">
    <Header size="small">Access denied:</Header>
    <p>
      Your account does not have permission to view the complete content of this
      tutorial. If you believe this is an error please contact support.
    </p>
  </Container>
);

export default TutorialAccessDenied;
