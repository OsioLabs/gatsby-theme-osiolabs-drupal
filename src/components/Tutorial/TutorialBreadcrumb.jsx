import React from 'react';
import PropTypes from 'prop-types';
import { Link } from 'gatsby';
import { Breadcrumb } from 'semantic-ui-react';

const TutorialBreadcrumb = props => (
  <Breadcrumb>
    <Breadcrumb.Section link as={Link} to="/">
      Home
    </Breadcrumb.Section>
    <Breadcrumb.Divider icon="right chevron" />
    <Breadcrumb.Section link as={Link} to={props.collection_path}>
      {props.collection_title}
    </Breadcrumb.Section>
    <Breadcrumb.Divider icon="right chevron" />
  </Breadcrumb>
);

TutorialBreadcrumb.propTypes = {
  collection_title: PropTypes.string,
  collection_path: PropTypes.string,
};

export default TutorialBreadcrumb;
