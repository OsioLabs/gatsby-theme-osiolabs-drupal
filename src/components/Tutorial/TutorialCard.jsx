import React from 'react';
import PropTypes from 'prop-types';
import { Link, navigate } from 'gatsby';
import { Card, Label, Icon } from 'semantic-ui-react';
import TutorialProgressIndicator from './TutorialProgressIndicator';

class TutorialCard extends React.Component {
  onClick = event => {
    event.preventDefault();
    const { path } = this.props;
    navigate(path);
  };

  render() {
    const {
      hasText,
      hasVideo,
      isPublic,
      isComingSoon,
      timeToComplete,
      title,
      path,
      summary,
      userAuthenticated,
    } = this.props;

    const displayLock = !userAuthenticated && !isPublic && !isComingSoon;

    return (
      <Card
        onClick={this.onClick}
        as="div"
        className={[
          isComingSoon && 'coming-soon',
          (isComingSoon || isPublic) && 'has-label',
        ].join(' ')}
      >
        {isPublic && !isComingSoon && (
          <Label as="span" corner="right" color="blue">
            Free!
          </Label>
        )}
        {isComingSoon && (
          <Label as="span" corner="right" color="red">
            Soon ...
          </Label>
        )}
        <Card.Content>
          <Card.Header as={Link} to={path}>
            {title}
            {' '}
            {displayLock && (
              <Icon
                name="lock"
                aria-label="Membership required. Preview available"
                size="small"
                color="red"
                fitted
              />
            )}
          </Card.Header>
          <Card.Description>
            {
              // Strip all the HTML tags, and then split out the first sentence.
              summary.replace(/<(?:.|\n)*?>/gm, '').split(/\.\s+/)[0]
              // The period after this is intentional because it's stripped above.
            }
            .
          </Card.Description>
        </Card.Content>
        <Card.Content>
          <Card.Meta className="left floated">
            {!userAuthenticated && (
              <TutorialProgressIndicator complete={false} entityId={1} />
            )}
            {hasText && (
              <Icon
                name="file text"
                aria-label="Tutorial contains text"
                size="small"
              />
            )}
            {hasVideo && (
              <Icon
                name="video play"
                aria-label="Tutorial has one or more videos"
                size="small"
              />
            )}
            {timeToComplete && ` ${timeToComplete}`}
          </Card.Meta>
          <Link to={path} className="right floated">
            {!displayLock && !isComingSoon ? 'Start now' : 'Preview'}
            <Icon name="angle double right" />
          </Link>
        </Card.Content>
      </Card>
    );
  }
}

TutorialCard.propTypes = {
  // Path to the tutorial relative to application root.
  path: PropTypes.string.isRequired,
  // Title of the tutorial.
  title: PropTypes.string.isRequired,
  // Summary of the tutorial.
  summary: PropTypes.string.isRequired,
  // Set to true if the tutorial is public and availble to anyone, otherwise set
  // to false if this is a members only tutorial.
  isPublic: PropTypes.bool,
  // Set to true if the tutorial is coming soon and should be displayed as a
  // teaser only.
  isComingSoon: PropTypes.bool,
  // Is the user viewing this authenticated?
  userAuthenticated: PropTypes.bool,
  // The tutorial contains text.
  hasText: PropTypes.bool,
  // The tutorial contains one more video embedded in it.
  hasVideo: PropTypes.bool,
  // 0-100, percent of the tutorial that the current viewer has completed.
  progress: PropTypes.number,
  // Time to complete. Estimated time in seconds it would take someone to
  // complete this tutorial.
  timeToComplete: PropTypes.number,
};

TutorialCard.defaultProps = {
  isPublic: false,
  isComingSoon: false,
  userAuthenticated: false,
  hasText: false,
  hasVideo: false,
  progress: 0,
  timeToComplete: null,
};

export default TutorialCard;
