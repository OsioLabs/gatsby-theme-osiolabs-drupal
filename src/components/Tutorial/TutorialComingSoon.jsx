import React from 'react';
import PropTypes from 'prop-types';
import Prism from 'prismjs';
import { Container, Header } from 'semantic-ui-react';

// Styling for code syntax highlighting.
require('prismjs/themes/prism.css');

const TutorialComingSoon = ({ title, summary }) => (
  <Container>
    <Header as="h1">{title}</Header>
    <Header as="h2">Coming soon:</Header>
    <div dangerouslySetInnerHTML={{ __html: summary }} />
  </Container>
);

TutorialComingSoon.componentDidMount = () => {
  // Whenever the component updates trigger Prism so that any new content will
  // get syntax highlighting.
  Prism.highlightAll();
};

TutorialComingSoon.propTypes = {
  title: PropTypes.string.isRequired,
  summary: PropTypes.string.isRequired,
  changed: PropTypes.instanceOf(Date).isRequired,
};

export default TutorialComingSoon;
