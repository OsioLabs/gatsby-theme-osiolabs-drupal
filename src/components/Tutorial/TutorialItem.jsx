import React from 'react';
import PropTypes from 'prop-types';
import { Link } from 'gatsby';
import { Item, Icon, Label } from 'semantic-ui-react';

const TutorialItem = ({
  hasText,
  hasVideo,
  path,
  progress,
  isPublic,
  sequence,
  timeToComplete,
  title,
}) => (
  <Item className="tutorial--item">
    {sequence && (
      <Item.Content>
        <Label circular>{sequence}</Label>
      </Item.Content>
    )}
    <Item.Content>
      <Item.Header>
        <Link to={path}>{title}</Link>
      </Item.Header>
      <Item.Meta>
        {hasText && (
          <Icon name="file text" aria-label="Tutorial contains text" />
        )}
        {hasVideo && (
          <Icon
            name="video play"
            aria-label="Tutorial has one or more videos"
          />
        )}
        {timeToComplete && ` ${timeToComplete}`}
      </Item.Meta>
    </Item.Content>
    {!isPublic && (
      <Item.Content>
        <Icon
          name="lock"
          aria-label="Membership required. Preview available"
          size="large"
        />
      </Item.Content>
    )}
  </Item>
);

TutorialItem.propTypes = {
  // Path to the tutorial relative to application root.
  path: PropTypes.string.isRequired,
  // Title of the tutorial.
  title: PropTypes.string.isRequired,
  // Set to true if the tutorial is public and available to anyone, otherwise
  // set to false if this is a members only tutorial.
  isPublic: PropTypes.bool,
  // If the tutorial is being displayed as part of a sequence this is it's
  // position in the list.
  sequence: PropTypes.number,
  // The tutorial contains text.
  hasText: PropTypes.bool,
  // The tutorial contains one more video embedded in it.
  hasVideo: PropTypes.bool,
  // 0-100, percent of the tutorial that the current viewer has completed.
  progress: PropTypes.number,
  // Time to complete. Estimated time in seconds it would take someone to
  // complete this tutorial.
  timeToComplete: PropTypes.string,
};

TutorialItem.defaultProps = {
  isPublic: false,
  sequence: null,
  hasText: false,
  hasVideo: false,
  progress: 0,
  timeToComplete: null,
};

export default TutorialItem;
