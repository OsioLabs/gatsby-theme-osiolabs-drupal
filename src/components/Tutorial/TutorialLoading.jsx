import React from 'react';
import PropTypes from 'prop-types';
import { Container, Placeholder } from 'semantic-ui-react';
import TutorialTeaser from './TutorialTeaser';

const TutorialLoading = props => (
  <Container>
    <TutorialTeaser {...props} />
    <p />
    <Placeholder fluid>
      <Placeholder.Header>
        <Placeholder.Line />
      </Placeholder.Header>
      <Placeholder.Paragraph>
        <Placeholder.Line />
        <Placeholder.Line />
        <Placeholder.Line />
        <Placeholder.Line />
        <Placeholder.Line />
      </Placeholder.Paragraph>
      <Placeholder.Header>
        <Placeholder.Line />
      </Placeholder.Header>
      <Placeholder.Paragraph>
        <Placeholder.Line />
        <Placeholder.Line />
        <Placeholder.Line />
      </Placeholder.Paragraph>
    </Placeholder>
  </Container>
);

TutorialLoading.propTypes = {
  title: PropTypes.string.isRequired,
  summary: PropTypes.string.isRequired,
};

export default TutorialLoading;
