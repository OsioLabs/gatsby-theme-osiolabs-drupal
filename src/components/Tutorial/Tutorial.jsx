import React from 'react';
import PropTypes from 'prop-types';
import { Container, Header, Label } from 'semantic-ui-react';

// Styling for code syntax highlighting.
import Prism from 'prismjs';
import TutorialAccessDenied from './TutorialAccessDenied';
import TutorialProgressIndicator from './TutorialProgressIndicator';

require('prismjs/themes/prism.css');

/**
 * Render a tutorial.
 *
 * Initially render a tutorial teaser using the <TutorialTeaser/> component,
 * with a placeholder graphic underneath. Then once the data has been loaded
 * from the server use that content instead.
 *
 * For use with <DrupalTutorial />.
 */
class Tutorial extends React.Component {
  componentDidMount() {
    Prism.highlightAll();
  }

  render() {
    const { error, title, summary, body, userAuthenticated } = this.props;

    return (
      <Container>
        {!userAuthenticated &&
          <TutorialProgressIndicator complete entityId={1} />
        }
        <Header as="h1">{title}</Header>
        {error && (
          // eslint-disable-next-line react/jsx-one-expression-per-line
          <Label basic color='red'>
            Unable to load tutorial content. {error.toString()}
          </Label>
        )}

        {summary && <div dangerouslySetInnerHTML={{ __html: summary }} />}
        <p />

        {body ? (
          <div dangerouslySetInnerHTML={{ __html: body }} />
        ) : (
          <TutorialAccessDenied />
        )}
      </Container>
    );
  }
}

Tutorial.propTypes = {
  id: PropTypes.string.isRequired,
  title: PropTypes.string.isRequired,
  summary: PropTypes.string.isRequired,
  body: PropTypes.string,
  changed: PropTypes.instanceOf(Date).isRequired,
  error: PropTypes.object,
};

export default Tutorial;
