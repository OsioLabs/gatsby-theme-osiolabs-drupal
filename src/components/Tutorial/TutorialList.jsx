import React from 'react';
import PropTypes from 'prop-types';
import { Item } from 'semantic-ui-react';
import TutorialItem from './TutorialItem';

const TutorialList = ({ tutorials }) => (
  <Item.Group>
    {tutorials.map((item, index) => (
      <TutorialItem
        title={item.title}
        path={item.path}
        sequence={index + 1}
        key={item.id}
        isPublic={item.tutorial_access === 'public'}
        isComingSoon={item.tutorial_access === 'coming_soon'}
      />
    ))}
  </Item.Group>
);

TutorialList.propTypes = {
  // Tutorials is an array of tutorial items with keys title, id, and path.
  tutorials: PropTypes.arrayOf(
    PropTypes.shape({
      id: PropTypes.string,
      title: PropTypes.string,
      path: PropTypes.string,
    })
  ).isRequired,
};

export default TutorialList;
