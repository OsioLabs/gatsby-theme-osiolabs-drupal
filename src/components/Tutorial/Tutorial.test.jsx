import React from 'react';
import { render } from '@testing-library/react';

import Tutorial from './Tutorial';

const props = {
  id: 'aaaa-bbbb-cccc-dddd',
  title: 'Tacos Rock',
  summary: '<p>Tacos <strong>go here</strong> ...</p>',
  changed: new Date(),
  processing: true,
};

describe('Component: <Tutorial />', () => {
  it('renders correctly', () => {
    const propsCopy = props;
    propsCopy.data = {
      attributes: {
        body: {
          processed: '<p>This is the full tutorial content.</p>',
        },
      },
    };
    const { container } = render(<Tutorial {...propsCopy} />);
    expect(container.firstChild).toMatchSnapshot();
  });

  // If there is no body content either in props.body or
  // props.data.attributes.body.processed show an access denied message.
  it('renders access deined', () => {
    const propsCopy = props;
    propsCopy.processing = false;
    const { container } = render(<Tutorial {...propsCopy} />);
    expect(container.firstChild).toMatchSnapshot();
  });

  it('renders an error', () => {
    const propsCopy = props;
    propsCopy.processing = false;
    propsCopy.error = new Error("Oh no, we're all out of cake.");
    const { container } = render(<Tutorial {...propsCopy} />);
    expect(container.firstChild).toMatchSnapshot();
  });
});
