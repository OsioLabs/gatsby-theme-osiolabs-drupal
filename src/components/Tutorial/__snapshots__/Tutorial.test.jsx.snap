// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Component: <Tutorial /> renders access deined 1`] = `
<div
  class="ui container"
>
  <h1
    class="ui header"
  >
    Tacos Rock
  </h1>
  <div>
    <p>
      Tacos 
      <strong>
        go here
      </strong>
       ...
    </p>
  </div>
  <p />
  <div
    class="ui container tutrial--access-denied"
  >
    <div
      class="ui small header"
    >
      Access denied:
    </div>
    <p>
      Your account does not have permission to view the complete content of this tutorial. If you believe this is an error please contact support.
    </p>
  </div>
</div>
`;

exports[`Component: <Tutorial /> renders an error 1`] = `
<div
  class="ui container"
>
  <h1
    class="ui header"
  >
    Tacos Rock
  </h1>
  <div
    class="ui red basic label"
  >
    Unable to load tutorial content. 
    Error: Oh no, we're all out of cake.
  </div>
  <div>
    <p>
      Tacos 
      <strong>
        go here
      </strong>
       ...
    </p>
  </div>
  <p />
  <div
    class="ui container tutrial--access-denied"
  >
    <div
      class="ui small header"
    >
      Access denied:
    </div>
    <p>
      Your account does not have permission to view the complete content of this tutorial. If you believe this is an error please contact support.
    </p>
  </div>
</div>
`;

exports[`Component: <Tutorial /> renders correctly 1`] = `
<div
  class="ui container"
>
  <h1
    class="ui header"
  >
    Tacos Rock
  </h1>
  <div>
    <p>
      Tacos 
      <strong>
        go here
      </strong>
       ...
    </p>
  </div>
  <p />
  <div
    class="ui container tutrial--access-denied"
  >
    <div
      class="ui small header"
    >
      Access denied:
    </div>
    <p>
      Your account does not have permission to view the complete content of this tutorial. If you believe this is an error please contact support.
    </p>
  </div>
</div>
`;
